examples/shadow_neighborloader.py

```diff
@@ -0,0 +1,94 @@
+import os.path as osp
+
+import torch
+import torch.nn.functional as F
+from tqdm import tqdm
+
+from torch_geometric.datasets import Flickr
+from torch_geometric.loader import NeighborLoader
+from torch_geometric.nn import SAGEConv
+
+path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'Flickr')
+dataset = Flickr(path)
+data = dataset[0]
+
+kwargs = {'batch_size': 1024, 'num_workers': 10, 'persistent_workers': True}
+
+# Mimic ShadowGNN by setting the argument subgraph_type='induced'.
+# This along with the num_neighbors argument will get the
+# complete induced subgraph
+train_loader = NeighborLoader(data, num_neighbors=[5] * 2,
+                              input_nodes=data.train_mask,
+                              subgraph_type='induced', **kwargs)
+val_loader = NeighborLoader(data, num_neighbors=[5] * 2,
+                            input_nodes=data.val_mask, subgraph_type='induced',
+                            **kwargs)
+test_loader = NeighborLoader(data, num_neighbors=[5] * 2,
+                             input_nodes=data.test_mask,
+                             subgraph_type='induced', **kwargs)
+
+
+class GNN(torch.nn.Module):
+    def __init__(self, in_channels, hidden_channels, out_channels):
+        super().__init__()
+        self.conv1 = SAGEConv(in_channels, hidden_channels)
+        self.conv2 = SAGEConv(hidden_channels, hidden_channels)
+        self.conv3 = SAGEConv(hidden_channels, hidden_channels)
```

Comment:
Just a bit confused - why do we use three convolutional layers when only two-hop neighbors are sampled?

Reply:
First, the sampling depth and model depth are independent parameters. Meaning, one can have a subgraph sampled using 2 hops and such a subgraph can be used on a model with 3 hops. The same node might appear on multiple levels as a result.

Second, the core feature of ShadowGNN is that is can sample shallow sub-graphs and enable deeper GNNS.